src/main/kotlin/edu/kit/compiler/Compiler.kt

```diff
@@ -49,125 +37,120 @@ class Compiler(private val config: Config) {
                     throw NotImplementedError("Compile mode not yet implemented.")
                 }
-                Mode.Echo -> runBlocking {
-                    withContext(CoroutineScope(threadPool!!.asCoroutineDispatcher()).coroutineContext) {
-                        try {
-                            var c = input.next()
-                            while (c != InputProvider.END_OF_FILE) {
-                                print(c)
-                                c = input.next()
-                            }
-                            return@withContext ExitCode.SUCCESS
-                        } catch (e: IOException) {
-                            System.err.println("Unexpected IOException: ${e.message}")
-                            return@withContext ExitCode.ERROR_FILE_SYSTEM
-                        }
-                    }
-                }
-                Mode.LexTest -> runBlocking {
-                    var hasInvalidToken = false
-
+                Mode.LexTest -> {
                     Lexer(
-                        config.sourceFile.absolutePath,
-                        input,
+                        sourceFile,
                         stringTable,
                         printWarnings = false
-                    ).tokens().onEach {
-                        hasInvalidToken = hasInvalidToken || (it is Token.ErrorToken)
-                    }.lexTestRepr.collect {
+                    ).tokens().lexTestRepr.forEach {
                         println(it)
                     }
 
-                    if (hasInvalidToken) {
-                        ExitCode.ERROR_INVALID_TOKEN
-                    } else {
-                        ExitCode.SUCCESS
+                    if (sourceFile.hasError) {
+                        return ExitCode.ERROR_COMPILATION_FAILED
                     }
                 }
-                Mode.ParseTest -> runBlocking {
-                    var hasInvalidToken = false
-
-                    val tokens = Lexer(
-                        config.sourceFile.absolutePath,
-                        input,
+                Mode.ParseTest -> {
+                    val lexer = Lexer(
+                        sourceFile,
                         stringTable,
                         printWarnings = false
-                    ).tokens()
-                    try {
-                        val program = Parser(tokens).parse()
-                        println(program)
-                    } catch (ex: IllegalArgumentException) {
-                        System.err.println("error: $ex")
-                    }
+                    )
+                    val parser = Parser(lexer.tokens())
+                    parser.parse()
 
-                    ExitCode.SUCCESS
+                    if (sourceFile.hasError) {
+                        return ExitCode.ERROR_COMPILATION_FAILED
+                    }
                 }
+                else -> throw IllegalStateException("unhandled compilation mode")
```

Comment:
das sollte nie passieren, oder? (wenn der mode nicht unterstützt wird wirft clikt einen fehler,oder?)

Reply:
Jo, is da, weil sonst der Echo case nicht gehandelt wird. Und da das nur ne warning ist, dachte ich lieber so^^

Reply:
warte.. warum wird der nicht gehandelt?

Reply:
das würde doch dann bei `--echo` mit "unhandled compilation mode" failen

Reply:
muss vllt nicht in diesem PR gelöst werden, aber ich fänd dann
```kotlin
when(this.mode) {
    is Mode.Echo -> handleEcho()
    is Mode.LexTest -> { val source = readSource(...); Lexer(source...)..... }
    is Mode.ParseTest -> { val source = readSource(...); Lexer(source,...) ...; Parser(..).parse() }
}
```
schöner, da das `when` dann exaustive ist

Reply:
ich schätze dafür braucht man dann auch einen `data class ConsoleError(val returnCode, val message) : ...` o.ä.

Reply:
Geht nicht, weil das SourceFile vorher erstellt wird und das SourceFile erwartet en valides ASCII file. Echo soll aber (warum auch immer) ja auch binary files unterstützen...